### ISSUE TITLE
Systemd / init.d autostart script

### ISSUE THREAD

**Issue by mzuenni** [airdcpp-web/airdcpp-webclient#152]
Hey another enhancement would be to automatically install and give the user the ability to activate a Systemd or/and init.d script.
In my time i want to start it at boot time (like an apachi server) i already tried to write an Systemd script my self but nothing happens...
```
[Unit]
Description = airdcppd daemon
After = NetworkManager-wait-online.service network.target network-online.target
Wants = display-manager.service NetworkManager-wait-online.service network-online.target

[Service]
Type = simple
PIDFile = /var/run/airdcppd.pid
ExecStart = /usr/local/bin/airdcppd -d

[Install]
WantedBy = multi-user.target
```

**Comment by acmodeu** [comment]
[This](https://github.com/airdcpp-web/airdcpp-webclient/issues/137#issuecomment-208759531) is what I did, maybe it will help

**Comment by maksis** [comment]
The pid file is also located inside the config directory.

https://github.com/airdcpp-web/airdcpp-webclient/issues/38 :wink:

**Comment by stevenmirabito** [comment]
This works for me:

```
[Unit]
Description=A peer-to-peer file sharing client with web user interface
After=network.target

[Service]
User=airdcpp
Group=media
ExecStart=/usr/local/bin/airdcppd -c=/var/lib/airdcpp

[Install]
WantedBy=multi-user.target
```

I'm running AirDC++ under its own user (airdcpp), so modify accordingly to suit your setup. Because the daemon doesn't fork or do anything weird, systemd will keep track of the process automatically.

**Comment by mzuenni** [comment]
hello i also tried your script but still it does nothing...
my script currently looks like this:
```
[Unit]
Description=airdcppd daemon
After=network.target

[Service]
User=zuenni
Group=users
ExecStart=/usr/local/bin/airdcppd -c="/media/data/airdcpp/config" -web-resources="/media/data/airdcpp/webUI"

[Install]
WantedBy=multi-user.target
```
and the file is located at "/etc/systemd/system/airdcppd.service"
then i run :
```
sudo systemctl enable airdcppd.service
sudo systemctl start airdcppd.service
```
to test it but the program doesnt start but i dont get any feedback....
(if i manually run the ExecStart line manually everything works like expected)

is there anything i should do? 
(its the first time i an using systemd)

**Comment by maksis** [comment]
I tried that script and it works fine for me. Are you still having problems?

**Comment by mzuenni** [comment]
yes it works now for me too. i dont know what caused the problem for me ^^

**Comment by KipourosV** [comment]
Normally I run airdcppd with the -d switch and nothing else.
How would I write the ExecStart= ??

Do I keep the -d switch?

**Comment by maksis** [comment]
Based on the previous comments, I wouldn't keep that switch

**Comment by KipourosV** [comment]
There is a guide on creating the system scripts for deluge on their site.
They also use the -d switch for deluge and have the switch in their scipt example...

**Comment by maksis** [comment]
It should be easy to test whether the option works. Have you tried to run the script? It will spam the typical console messages to syslog when not using the -d option but there are other ways to disable such output if needed.

**Comment by maksis** [comment]
You may get strange errors when the application is launched on system startup before network connectivity is available.

Based on [this article](https://www.freedesktop.org/wiki/Software/systemd/NetworkTarget/), `After=network.target` is almost meaningless.

I replaced `After=network.target` with `After=network-online.target` and based on a few reboots, it seemed to have fixed the issue for me.

**Comment by piratthefan** [comment]
Hi i need abit of gelp i cant get this to auto start.
This works great

sudo systemctl enable airdcppd.service
sudo systemctl start airdcppd.service 

But it wont start even if i change to After=network-online.target

**Comment by maksis** [comment]
Could you paste the full script that you are using?

**Comment by maksis** [comment]
Can you find anything useful from `/var/log/syslog`?

**Comment by piratthefan** [comment]
i got this that where red but dont think it have something to whit it  

Apr 17 15:35:59 share kernel: [    0.000000]    Build-time adjustment of leaf fanout to 64.
Apr 17 15:35:59 share kernel: [    0.000000]    RCU restricting CPUs from NR_CPUS=512 to nr_cpu_ids$

also Apr 17 15:36:05 share airdcppd[1121]: Starting.
Apr 17 15:36:05 share airdcppd[1121]: Failed to create PID file /home/airdcpp/.airdc++/airdcppd.pid$
Apr 17 15:36:05 share systemd[1]: airdcppd.service: Main process exited, code=exited, status=1/FAIL$
Apr 17 15:36:05 share systemd[1]: airdcppd.service: Unit entered failed state.
Apr 17 15:36:05 share systemd[1]: airdcppd.service: Failed with result 'exit-code'.
Apr 17 15:36:06 share iscsid: iSCSI daemon with pid=1131 started!
Apr 17 15:36:08 share systemd[1]: Created slice User Slice of airdcpp.
Apr 17 15:36:08 share systemd[1]: Starting User Manager for UID 1001...
Apr 17 15:36:08 share systemd[1]: Started Session 1 of user airdcpp.

**Comment by maksis** [comment]
```Apr 17 15:36:05 share airdcppd[1121]: Failed to create PID file /home/airdcpp/.airdc++/airdcppd.pid$```

That very much explains it. You should check the permissions of the directory `/home/airdcpp/.airdc++/` and its content (is the directory owned by a different user?)

**Comment by piratthefan** [comment]
hmm det owner is airdcpp but should it be that i ecryptfs the home dir :P

**Comment by maksis** [comment]
Yeah, you should put the correct config directory there. Please see `airdcppd --help` for information about specifying a custom config directory.

**Comment by maksis** [comment]
Are you able to try running the app as the other user without using systemd to see if it works?

**Comment by piratthefan** [comment]
yep systemctl start, stop and status works thats why i dont get it.
Dont have other path to config. Been looking for it in the airdc++ folder and web.

**Comment by maksis** [comment]
> yep systemctl start, stop and status works thats why i dont get it.

systemctl won't detect such permission issues. What's the output of `ls -la /home/airdcpp/.airdc++/`?

**Comment by piratthefan** [comment]
drwxrwxr-x 10 airdcpp airdcpp    4096 Apr 17 17:35 .
drwx------  7 airdcpp airdcpp    4096 Apr 17 15:37 ..
-rw-r--r--  1 airdcpp shareDC     101 Apr 17 15:35 ADLSearch.xml
drwxrwxr-x  2 airdcpp airdcpp    4096 Apr 15 22:03 Bundles
drwxrwxr-x  2 airdcpp airdcpp    4096 Apr 15 22:03 Certificates
-rw-r--r--  1 airdcpp shareDC    3174 Apr 17 15:35 DCPlusPlus.xml
-rw-r--r--  1 airdcpp shareDC    1183 Apr 17 13:32 Favorites.xml
-rw-r--r--  1 airdcpp airdcpp    1183 Apr 17 16:48 Favorites.xml.bak
drwxrwxr-x  2 airdcpp airdcpp    4096 Apr 17 16:48 FileIndex
drwxrwxr-x  2 airdcpp airdcpp    4096 Apr 15 22:03 FileLists
-rw-rw-r--  1 airdcpp airdcpp 1082230 Apr 15 22:04 GeoIP.dat
-rw-rw-r--  1 airdcpp airdcpp  620142 Apr 15 22:04 GeoIP.dat.gz
-rw-rw-r--  1 airdcpp airdcpp 1981979 Apr 15 22:04 GeoIPv6.dat
-rw-rw-r--  1 airdcpp airdcpp  984800 Apr 15 22:04 GeoIPv6.dat.gz
drwxrwxr-x  2 airdcpp airdcpp   32768 Apr 17 17:35 HashData
drwxrwxr-x  2 airdcpp airdcpp    4096 Apr 15 22:03 HubLists
drwxrwxr-x  3 airdcpp airdcpp    4096 Apr 17 00:11 Logs
-rw-rw-r--  1 airdcpp airdcpp     625 Apr 16 17:16 Recents.xml
-rw-r--r--  1 airdcpp airdcpp       0 Apr 17 16:48 RUNNING
drwxrwxr-x  2 airdcpp airdcpp    4096 Apr 17 16:49 ShareCache
-rw-r--r--  1 airdcpp shareDC     317 Apr 17 15:35 WebServer.xml

**Comment by maksis** [comment]
Can run the app with the wanted user and group from terminal by using the [`runuser`](http://man7.org/linux/man-pages/man1/runuser.1.html) command?

**Comment by maksis** [comment]
```Apr 17 15:36:05 share airdcppd[1121]: Failed to create PID file /home/airdcpp/.airdc++/airdcppd.pid$```

There's also more text on the right, that will show the exact error message.

**Comment by piratthefan** [comment]
Apr 17 16:06:33 share airdcppd[1107]: Failed to create PID file /home/airdcpp/.airdc++/airdcppd.pid: No such file or directory

but airdcpp is a root user 

dont get the runuser command 

airdcpp@share:/var/log$ runuser -l airdcpp -c airdcpp
runuser: may not be used by non-root users

**Comment by piratthefan** [comment]
Or is it dum too have it as a root user?

**Comment by maksis** [comment]
It means that you are expected to run the command as root

**Comment by piratthefan** [comment]
Try to run $sudo systemctl start airdcppd.service 
from another use but i got errors

airdcppd.service - A peer-to-peer file sharing client with web user interface
Loaded: loaded (/etc/systemd/system/airdcppd.service; disabled; vendor preset: enabled)
Active: failed (Result: exit-code) since Mon 2017-04-17 22:33:20 CEST; 1min 49s ago
Process: 1388 ExecStart=/usr/local/bin/airdcppd (code=exited, status=1/FAILURE)
Main PID: 1388 (code=exited, status=1/FAILURE)

If it helps

**Comment by piratthefan** [comment]
Hi i got it to work. I think the problem was that i encrypted the home dir. :P Thx for the help

**Comment by Kcchouette** [comment]
I'm doing a autostart using screen `su --command="screen -dmS ${user}-airdcpp airdcpp" "${user}"` and a init.d script I've put in `/etc/init.d/`

**Comment by KipourosV** [comment]
[Unit]
Description=airdcppd daemon
After=network-online.target

[Service]
Type=simple
User=airdcppd
Group=media
UMask=007
PIDFile = /root/.airdc++/airdcppd.pid
ExecStart = /usr/local/bin/airdcppd

Restart=on-failure

# Configures the time to wait before service is stopped forcefully.
TimeoutStopSec=300

[Install]
WantedBy=multi-user.target



This works for me. 

FYI:
The umask is the mask for permissions on files created by airdcpp (new downloads)
007 means the created files get 770.

**Comment by cirojr82** [comment]
I'm not SO pro as you guys are...I just set it in CRONTAB to after every boot run airdc.

**Comment by riptide981** [comment]
[Unit]
Description = A peer-to-peer file sharing client with web user interface
After = network-online.target

[Service]
Type = simple
User = p2p
Group = users
UMask = 007
PIDFile = /home/p2p/.config/AirDC++/airdcppd.pid
ExecStart = /home/p2p/airdcpp-webclient/airdcppd -d -p="/home/p2p/.config/AirDC++/airdcppd.pid" -c="/home/p2p/.config/AirDC++/" -web-resources="/home/p2p/airdcpp-webclient/web-resources"

[Install]
WantedBy = multi-user.target

**Comment by cam8001** [comment]
Slightly modified from above. I had to disable selinux on my Fedora 32 Server distro.

```ini
[Unit]
Description = A peer-to-peer file sharing client with web user interface
After = network-online.target

[Service]
Type = simple
User = airdcpp
Group = airdcpp
UMask = 007
PIDFile = /home/airdcpp/bin/airdcppd.pid
ExecStart = /usr/local/bin/airdcppd-webclient/airdcppd -p="/home/airdcpp/bin/airdcppd.pid" -c="/home/airdcpp/.config/airdcppd/config" -web-resources="/usr/local/bin/airdcppd-webclient/web-resources"

[Install]
WantedBy = multi-user.target
```

**Comment by gmipf** [comment]
Here is how I did it: 

```
[Unit]
Description=A peer-to-peer file sharing client with web user interface
After=network.target
ConditionUser=!root   #Skips the root user.

[Service]
ExecStartPre=/bin/sh -c 'until ping -c1 google.com; do sleep 1; done;'   #Wait for internet connection.
ExecStart=/usr/local/bin/airdcppd
Restart=on-failure
Type=simple

[Install]
WantedBy=default.target

```
My distro is Kubuntu 20.04. i compiled the binary from source and installed to default install location using `sudo make install`.

I have created it the service as a user service which does not need any root permissions. User level services needs to be created in a different location:
`/usr/lib/systemd/user/airdcppd.service`

To control user services you need to add the `--user` parameter, e.g.:
```
systemctl --user start airdcppd.service
systemctl --user status airdcppd.service
```
And it does not need sudo or doesn't even work with sudo in my case. 

The config files are saved at the default location `/home/$HOME/.airdc++/` 

As reference I have used a combination of the above and the pulseaudio user service. 

I had to add the following otherwise at boot the client would start before it has internet connection and would give connection errors and also would change the client bind address from my local client ip back to `0.0.0.0`, which I don't want.
`ExecStartPre=/bin/sh -c 'until ping -c1 google.com; do sleep 1; done;' `

PS: Setting the bind address to `localhost` isn't possible, which is weird. So instead I had to set the bind address to my current local ip.
